### ISSUE TITLE
Ability to create user functions in process variables and pass between process contexts

### ISSUE THREAD

**Issue by jphillips-bp3** [camunda/camunda#20141]
**Is your feature request related to a problem? Please describe.**
We have a situation with a customer that have a lot of repeated logic defined in FEEL expressions that is used in multiple places across their process. The issue here is that if this logic needs to change, then it has to be changed in multiple places.

**Describe the solution you'd like**
We would like to be able to define user defined functions within a process variable that could be then be within a process context, or indeed passed into sub-process via input mappings to be reused for example like any other process variable.

**Describe alternatives you've considered**
It is not practical to do the following:
- Create a common process fragment with a script task that defines a FEEL expression that can be called via a Call Activity. This would bloat the process out with a lot of steps than it already has. These functions are not really business focused and should not pollute the process model.
- Defer these to an external worker or connector, because again like above, we would have to create extra process steps to execute the logic. Plus developing and hosting a service for what would be relatively small helper functions would be overkill.

**Additional context**
This is what we would like to be able to define, but within in process variables: https://docs.camunda.io/docs/components/modeler/feel/language-guide/feel-functions/#user-defined

**Comment by jphillips-bp3** [comment]
@aleksander-dytko Hi, any movement on this at all? We have a C7 to C8 migration coming up (several in fact) where we have common Groovy scripts we can convert to FEEL. But we would like to be able to define custom functions and reuse and not have to develop workers. This would be a huge benefit to us. Thanks in advance.

**Comment by aleksander-dytko** [comment]
@Balazs-PM could you take a look at this request? I believe it would be a great addition in our modeling environments.

**Comment by saig0** [comment]
@aleksander-dytko / @Balazs-PM in the discovery phase, we should explore the best option to share a custom FEEL function. The user proposed to store the function in a variable. An alternative could be a globally-defined function, similar to cluster variables. This could be useful for BPMN as well as DMN. I saw similar requests in the past. Technically, this is a core feature of the engine because the engine doesn't support this yet.

**Comment by jphillips-bp3** [comment]
> [@aleksander-dytko](https://github.com/aleksander-dytko) / [@Balazs-PM](https://github.com/Balazs-PM) in the discovery phase, we should explore the best option to share a custom FEEL function. The user proposed to store the function in a variable. An alternative could be a globally-defined function, similar to cluster variables. This could be useful for BPMN as well as DMN. I saw similar requests in the past. Technically, this is a core feature of the engine because the engine doesn't support this yet.

A globally defined function would be even better, so we would have reuse across disconnected processes, much like connector templates are defined.

**Comment by Balazs-PM** [comment]
Would this be a functionality on the cluster side, where (CoE) or other privileged users can maintain these, and modeler users can refer to as they would do with  _stored procedures_? 

fyi: @nikku

**Comment by saig0** [comment]
@Balazs-PM yes. If we follow this idea, we need to add an API to deploy/store FEEL functions, maybe similar to [cluster variables](https://docs.camunda.io/docs/next/apis-tools/orchestration-cluster-api-rest/specifications/create-global-cluster-variable/) or as a new deployment resource type. From the Modeler side, we need to enable users to write and publish the FEEL functions. For a nice UX, we might fetch the available FEEL functions from the cluster to feed them into the FEEL expression editor for auto-completion and discovery.

**Comment by nikku** [comment]
For context: This has come up in different ways several times in the past - and we discussed different approaches:

* Defined on the cluster, such as cluster variables - similar to database "stored procedures" maybe
* Deployed as resource, such as a BPMN process, re-usable across different places - similar to, a BPMN process, with a similar deployment and update model
* Defined in a BPMN diagram, as part of a BPMN process, and re-usable "in local scope"

---

From the modeler there has to be a way to learn about such functions, so we can extend our editor intelligence to them.
